### PR TITLE
LibJS+Lagom: Add a Clang tool to enforce proper use of GCPtr/NonnullGCPtr

### LINKED ISSUE #16988: LibJS: Verify that GCPtr and NonnullGCPtr are only used on GC'd types
The trivial fix for this is adding either a static_assert or a requires clause to the class declarations/definitions for the fancy pointer types.

However, this doesn't work when using ``AK::IsBaseOf<JS::Cell, T>``. IsBaseOf is implemented in terms of the `__is_base_of` intrinsic, which can only be used with complete types. There are dozens of cases in LibJS and LibWeb where we either want to return or hold onto a GCPtr of a type that is only forward declared. Requiring the full definition of these types in those cases is a non-starter for build times. Additionally, any type U that holds a GCPtr<U> as a member (e.g. parent, child, sibling, ...) will fail to compile as the type is not deemed complete per the trait's requirements until the end of the class declaration.

Options:
- Continue to not verify this precondition of the type ( :disappointed:  )
- Create a bespoke python or bash check ( :fearful: )
- Create a clang-tidy check
- Create a compiler plugin to verify the condition after each source file is parsed (GCC? Clang? Both?)
- Come up with a cleverer meta-programming technique than checking that T has JS::Cell as an ancestor class.

### PR DESCRIPTION
**Pull request by mattco98**
The tool currently validates two things:
  - For all fields of type `GCPtr<T>` or `NonnullGCPtr<T>`, that `T` inherits from `JS::Cell`
  - For all fields of type `T&` or `T*` not wrapped in `GCPtr` or `NonnullGCPtr`, that `T` does not inherit from `JS::Cell` (otherwise it should be wrapped in a `GCPtr`/`NonnullGCPtr`)

This PR does not actually do any field conversions, since there are a _lot_ of unwrapped Cells around the codebase and it'd be too much for this PR. However, in order to prepare for that change, I've ported the const laundering changes from (NN)RefPtr (3c7a0ef1ac279c99bc4ee72085278ea70c210889) to (NN)GCPtr so we can properly propogate `const`. I've also fixed a bunch of compilation issues that caused the tool to error out. I'm not sure why they don't appear with a normal clang build of Serenity, but they seem like valid enough errors. 

A few notes:
  - If anyone demos it locally, I would suggest commenting out the `validate_non_gcptr_field` call in `CellHandler.cpp` as that produces a ton of output. 
  - I tried to integrate it into the Lagom build system (i.e. using `lagom_tool`) so that it could be invoked from `Meta/serenity.sh` like any other tool, but I couldn't figure out how to multithread the invocation like the python wrapper script currently does (I was getting scary internal malloc errors). So I figured this could be left until later if anyone cares to convert it. 

There are also a few errors that the tool produces:
  - [This error](https://i.imgur.com/n7TOZeK.png), not sure why it happens or what to do about it
  - The dollar sign in `$262Object.h` causes issues ([error](https://i.imgur.com/Y46UXnP.png)). I tested removing the dollar sign and it fixes it. Not sure if that's a change we're fine making for the sake of this tool, but if it's fine I'll go ahead and make that change. 
  - There is also an incomplete type error between `Web::CSS::CalculatedStyleValue` and `Web::CSS::StyleValue`. I was thinking about moving `CalculatedStyleValue` into its own file, but not sure if that's possible/if it'd be acceptable.

Closes #16988

### CODE REVIEW

File: Meta/Lagom/Tools/LibJSGCVerifier/src/main.py
Comment (linusg) [inline Meta/Lagom/Tools/LibJSGCVerifier/src/main.py:5]: `os.path` is banned in modern Python, use `pathlib`

File: Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.h
Comment (linusg) [inline Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.h:9]: looks weird, please follow the usual style please:

https://github.com/SerenityOS/serenity/blob/7c1455c3576c11f08ea7fc638b5106a11f9a7d0f/Userland/Libraries/LibWeb/Bindings/PlatformObject.h#L26-L28

**Review by linusg — CHANGES_REQUESTED** [review]
Some comments on commit prefixes:

- `Services/WebContent:` -> `WebContent:`
- `LibWeb/DOM+HTML` -> `LibWeb`, touches more than just DOM and HTML anyway. Can also be squashed with the following commit, I don't see a reason for them to be standalone.

**Review by linusg — COMMENTED** [review]
Regarding this:

> The dollar sign in $262Object.h causes issues ([error](https://i.imgur.com/Y46UXnP.png)). I tested removing the dollar sign and it fixes it. Not sure if that's a change we're fine making for the sake of this tool, but if it's fine I'll go ahead and make that change.

I'd rather not change a perfectly valid filename that matches the object's name to avoid a bug. Whatever is altering the path to turn one dollar sign into two should just not do that instead :)

**Comment by mattco98** [comment]
> I'd rather not change a perfectly valid filename that matches the object's name to avoid a bug. Whatever is altering the path to turn one dollar sign into two should just not do that instead :)

Yeah I figured. The file looks normal in both `main.py` and the `compile_commands.json` file, so maybe it's a clang tooling issue? Not totally sure, but I think I'll just leave it for now.

File: Meta/Lagom/Tools/LibJSGCVerifier/.gitignore
Comment (ADKaster) [inline Meta/Lagom/Tools/LibJSGCVerifier/.gitignore:1]: From the PR description: 
> [This error](https://i.imgur.com/n7TOZeK.png), not sure why it happens or what to do about it

This looks like has to do with include path handling in the main build. If you insert an include path that has standard libC header files *before* the compiler default include paths in the search list, all the `#include_next` behavior gets super messed up in clang/libc++'s builtin header files that patch over LibC header files to make them wrappable by the C++ STL. We talked about this on... the syzcaller issue iirc.

**Review by ADKaster — COMMENTED** [review]
Very cool! I have a couple of comments. One thing that might be nice is moving the "usage" instructions from the PR description into a README.md in the tool folder itself.

File: Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp
Comment (mattco98) [reply Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp:85]: I think you're right yeah, that case isn't caught. I did a test that was all in the same file and it worked fine, but I guess when you split the file up it no longer works. I'm not totally sure how to fix this at the moment, it doesn't work even if I give clang both file paths to look up. I was under the impression that they'd be resolved at the point `::run` is ran, but apparently not. I'll keep looking into it.

Comment (ADKaster) [reply Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp:85]: @nico do you have any ideas on how to catch this problem using clang's C++ API? I saw your name in the git history for chromium's similar tools (and I think you linked me to them before :) ).

My understanding is that your tool's `run()` will be called when an AST match occurs, and that the CompilerInstance/ASTContext is only extant until `handleEndSource` is called on the tool (which you didn't override :)). 

So currently, the tool only looks at `clang::CXXRecordDecl`s that are themselves `JS::Cell`s, and checks their own members for errant pointers. But if the data members of the classes that you're checking are pointers to incomplete types (as we tend to do .. :) ). Then we won't know about them. I think that only checking `.cpp` files might help, as presumably the classes we're using would need their complete definition in order to allocate them on the JS Heap, but that might not be true.

Comment (ADKaster) [reply Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp:85]: Doing the analysis in the `handleEndSource` callback might also help? As at that point, we can ask the QualType for their canonicalDecl or whatever the correct API is to get the actual definition of each type before checking its inheritance tree.

Comment (mattco98) [reply Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp:85]: I'm pretty sure the solution here is to remove the optimization where I only visit each class once. This way we'll visit a class every time it's used (i.e. from a cpp file), and it is effectively guaranteed that it will have a complete definition in at least one of them. I tested by adding a `struct Foo;` and `GCPtr<Foo> m_dummy` to `StringIterator.h` and a corresponding `struct Foo {};` to `StringIterator.cpp` and it was detected, so I think we're good here. While I'm at it I'll implement your suggestion to only look at .cpp files, since the headers don't really do anything for us.

File: Meta/Lagom/Tools/LibJSGCVerifier/src/main.py
Comment (mattco98) [reply Meta/Lagom/Tools/LibJSGCVerifier/src/main.py:49]: Looks like it isn't required anymore. I think I needed this before I set up my CMakeLists.txt properly.

**Comment by mattco98** [comment]
So it looks like the dollar sign issue happens with whatever exports the `compile_commands.json` (the closest thing I could find was `llvm-project/clang/lib/Driver/ToolChains/Clang.cpp:2423`, but that output looks different so I doubt that's where it happens). The entry the Clang generates for `$262Object` is:
```json

{
  "directory": "/home/matthew/code/serenity/Build/x86_64clang",
  "command": "/home/matthew/code/serenity/Toolchain/Local/clang//bin/clang++ /*bunch of options here*/ -o Userland/Libraries/LibJS/CMakeFiles/LibJS.dir/Contrib/Test262/$262Object.cpp.o -c \"/home/matthew/code/serenity/Userland/Libraries/LibJS/Contrib/Test262/\\$$262Object.cpp\"",
  "file": "/home/matthew/code/serenity/Userland/Libraries/LibJS/Contrib/Test262/$262Object.cpp"
},
```
So the issue lies with whatever is doing the escaping for that `-c` option. I did manage to come up with a patch to fix it on the reading side though:
```patch
diff --git a/clang/lib/Tooling/JSONCompilationDatabase.cpp b/clang/lib/Tooling/JSONCompilationDatabase.cpp
index 5e18d7a576c0..22d8f7b76b61 100644
--- a/clang/lib/Tooling/JSONCompilationDatabase.cpp
+++ b/clang/lib/Tooling/JSONCompilationDatabase.cpp
@@ -110,7 +110,12 @@ class CommandLineArgumentParser {
 
   bool skipEscapeCharacter() {
     if (*Position == '\\') {
-      return next();
+      if (next()) {
+      	if (*Position == '$')
+      	  return next();
+     	return true;
+      }
+      return false;
     }
     return true;
   }
```
Doesn't seem worth adding, especially since `$262Object.h` is quite small relative to other JS-related files, but at the very least it was fun to poke around the clang codebase for this.

**Comment by ADKaster** [comment]
The compile commands are created by CMake fwiw

**Comment by mattco98** [comment]
Ahh ok, good to know. I think for now I'll just ignore that file in the python script and add a FIXME for it.

**Comment by ADKaster** [comment]
<https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/cmNinjaTargetGenerator.cxx#L1748> and <https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/cmGlobalNinjaGenerator.cxx#L1169> seem the most relevant.

We can open an issue on their gitlab, but it might be that other consumers of compile_commands.json will do this escaping manually themselves. If clangd also chokes on the file, it might be on json producers to fix though.

File: Meta/Lagom/Tools/LibJSGCVerifier/.gitignore
Comment (mattco98) [reply Meta/Lagom/Tools/LibJSGCVerifier/.gitignore:1]: Yeah I don't think I'm knowledgeable enough to fix this :) at the end of the day it's a few line of error output, and I'd much rather spend my time using this tool to start the conversion rather than focus on this

**Comment by mattco98** [comment]
Realized I was only looking at inner types (i.e. the `B` in `A<B>`) for the invalid specialization diagnostic, so just pushed a change to apply that to both. It also makes the code look a bit nicer imo. Additionally, I've finished the next PR where I actually wrap everything, so the tool feels pretty robust at this point.

**Review by ADKaster — APPROVED** [review]
LGTM but I'll let Linus hit the merge button in case this affects other WIP LibJS work that I'm not aware of.

**Comment by ADKaster** [comment]
Looks like this accumulated some conflicts (Userland/Services/WebContent/WebDriverConnection.h)

**Comment by mattco98** [comment]
> Looks like this accumulated some conflicts (Userland/Services/WebContent/WebDriverConnection.h)

Fixed, thanks!

File: Userland/Libraries/LibJS/Heap/Handle.h
Comment (linusg) [inline Userland/Libraries/LibJS/Heap/Handle.h:49]: There's a handful of `make_handle(const_cast<T>(...))`s already to work around this, can we clean those up in the same commit?

Comment (mattco98) [reply Userland/Libraries/LibJS/Heap/Handle.h:49]: Good idea, cleaned!

File: Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.h
Comment (mattco98) [reply Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.h]: I agree. Though looking at it again, it looks like I can propogate the const-ness the other way (i.e. make `FetchParams::m_algorithms` const). Seems preferable to me, unless you think there's a reason to pass by value.

Comment (linusg) [reply Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.h]: yeah, that's fine. I love that `GCPtr` can now indicate constness!

**Pull request merged**
